File: php-checks/src/test/resources/checks/security/ChangingAccessibilityCheck.php
<?php

class MyClass
{
    public static $publicstatic = 'Static';
    private static $privatestatic = 'private Static';
    private $private = 'Private';
    private const CONST_PRIVATE = 'Private CONST';
    public $myfield = 42;

    private function __construct() {}
    private function privateMethod() {}
    public function __set($property, $value)  {}
    public function __get($property) {}
}

(new ReflectionClass('MyClass'))->getStaticProperties(); // Noncompliant
(new reflectionclass('MyClass'))->getstaticproperties(); // Noncompliant

$clazz = new ReflectionClass('MyClass');

$clazz->getStaticProperties(); // Noncompliant {{Make sure that this accessibility bypass is safe here.}}
$clazz->getSTATICPROPERTIES(); // Noncompliant

$clazz->setStaticPropertyValue('publicstatic', '42'); // OK as there is no overloading to bypass and it respects access control.
$clazz->getStaticPropertyValue('publicstatic'); // OK as there is no overloading to bypass and it respects access control.

// The following calls can access private or protected constants.
$clazz->getConstant('CONST_PRIVATE'); // Noncompliant
$clazz->getConstants(); // Noncompliant
$clazz->getReflectionConstant('CONST_PRIVATE'); // Noncompliant
$clazz->getReflectionConstants(); // Noncompliant

$obj = $clazz->newInstanceWithoutConstructor(); // Noncompliant

$constructorClosure = (new ReflectionClass('MyClass'))->getConstructor()->getClosure($obj); // Noncompliant
$constructorClosure = $clazz->getConstructor()->getClosure($obj); // Noncompliant
$constructor = $clazz->getConstructor();
$constructorClosure = $constructor->getClosure($obj); // Noncompliant
$constructor->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}

$prop = new \ReflectionProperty('MyClass', 'private');
$prop->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}
$prop->setValue($obj, "newValue"); // Noncompliant {{Make sure that this accessibility bypass is safe here.}}
$prop->getValue($obj); // Noncompliant

$prop2 = $clazz->getProperties()[2];
$prop2->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}
$prop2->setValue($obj, "newValue"); // Noncompliant
$prop2->getValue($obj); // Noncompliant

$meth = new ReflectionMethod('MyClass', 'privateMethod');
$clos = $meth->getClosure($obj); // Noncompliant
$meth->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}

$meth2 = $clazz->getMethods()[0];
$clos2 = $meth2->getClosure($obj); // Noncompliant
$meth2->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}

// Using a ReflectionObject instead of the class

$objr = new ReflectionObject($obj);
$objr->newInstanceWithoutConstructor(); // Noncompliant

$objr->getStaticPropertyValue("publicstatic"); // OK as there is no overloading to bypass and it respects access control.
$objr->setStaticPropertyValue("publicstatic", "newValue"); // OK as there is no overloading to bypass and it respects access control.

$objr->getStaticProperties(); // Noncompliant

// The following calls can access private or protected constants.
$objr->getConstant('CONST_PRIVATE'); // Noncompliant
$objr->getConstants(); // Noncompliant
$objr->getReflectionConstant('CONST_PRIVATE'); // Noncompliant
$objr->getReflectionConstants(); // Noncompliant

$constructor = $objr->getConstructor();
$constructorClosure = $constructor->getClosure($obj); // Noncompliant
$constructor->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}

$prop3 = $objr->getProperty('private');
$prop3->setAccessible(true); // Noncompliant
$prop3->setValue($obj, "newValue"); // Noncompliant
$prop3->getValue($obj); // Noncompliant

$prop4 = $objr->getProperties()[2];
$prop4->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}
$prop4->setValue($obj, "newValue"); // Noncompliant
$prop4->getValue($obj); // Noncompliant

$meth3 = $objr->getMethod('privateMethod');
$clos3 = $meth3->getClosure($obj); // Noncompliant
$meth3->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}

$meths = $objr->getMethods();
$meth4 = $meths[0];
$clos4 = $meth4->getClosure($obj); // Noncompliant
$meth4->setAccessible(true); // Noncompliant {{Make sure that this accessibility update is safe here.}}

foreach ($meths as $meth5){
  $meth5->getClosure($obj); // Noncompliant
}

foreach ($objr->getMethods() as $meth6){
  $meth6->getClosure($obj); // Noncompliant
}

foreach ($objr->getMethods() as $key => $meth7){
  $key->getClosure($obj);
  $meth7->getClosure($obj); // Noncompliant
}

File: php-checks/src/main/java/org/sonar/php/checks/security/ChangingAccessibilityCheck.java
/*
 * SonarQube PHP Plugin
 * Copyright (C) 2010-2025 SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.php.checks.security;

import java.util.function.Predicate;
import org.sonar.check.Rule;
import org.sonar.php.checks.utils.type.ArrayAccess;
import org.sonar.php.checks.utils.type.NewObjectCall;
import org.sonar.php.checks.utils.type.ObjectMemberFunctionCall;
import org.sonar.php.checks.utils.type.TreeValues;
import org.sonar.php.checks.utils.type.TypePredicateList;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;

@Rule(key = "S3011")
public class ChangingAccessibilityCheck extends PHPVisitorCheck {

  private static final String UPDATE_MESSAGE = "Make sure that this accessibility update is safe here.";
  private static final String BYPASS_MESSAGE = "Make sure that this accessibility bypass is safe here.";

  private static final NewObjectCall REFLECTION_CLASS = new NewObjectCall("ReflectionClass");
  private static final NewObjectCall REFLECTION_OBJECT = new NewObjectCall("ReflectionObject");
  private static final NewObjectCall REFLECTION_METHOD = new NewObjectCall("ReflectionMethod");
  private static final NewObjectCall REFLECTION_PROPERTY = new NewObjectCall("ReflectionProperty");

  private static final ObjectMemberFunctionCall REFLECTION_CLASS_GET_PROPERTY = new ObjectMemberFunctionCall("getProperty", REFLECTION_CLASS, REFLECTION_OBJECT);
  private static final ObjectMemberFunctionCall REFLECTION_CLASS_GET_PROPERTIES = new ObjectMemberFunctionCall("getProperties", REFLECTION_CLASS, REFLECTION_OBJECT);
  private static final ObjectMemberFunctionCall REFLECTION_CLASS_GET_METHODS = new ObjectMemberFunctionCall("getMethods", REFLECTION_CLASS, REFLECTION_OBJECT);
  private static final ObjectMemberFunctionCall REFLECTION_CLASS_GET_METHOD = new ObjectMemberFunctionCall("getMethod", REFLECTION_CLASS, REFLECTION_OBJECT);
  private static final ObjectMemberFunctionCall REFLECTION_CLASS_GET_CONSTRUCTOR = new ObjectMemberFunctionCall("getConstructor", REFLECTION_CLASS, REFLECTION_OBJECT);

  private static final Predicate<TreeValues> BYPASS_PREDICATES = new TypePredicateList(
    new ObjectMemberFunctionCall("getClosure",
      new ArrayAccess(REFLECTION_CLASS_GET_METHODS),
      REFLECTION_CLASS_GET_CONSTRUCTOR,
      REFLECTION_CLASS_GET_METHOD,
      REFLECTION_METHOD),

    new ObjectMemberFunctionCall("getConstant", REFLECTION_CLASS, REFLECTION_OBJECT),

    new ObjectMemberFunctionCall("getConstants", REFLECTION_CLASS, REFLECTION_OBJECT),

    new ObjectMemberFunctionCall("getReflectionConstant", REFLECTION_CLASS, REFLECTION_OBJECT),

    new ObjectMemberFunctionCall("getReflectionConstants", REFLECTION_CLASS, REFLECTION_OBJECT),

    new ObjectMemberFunctionCall("getStaticProperties", REFLECTION_CLASS, REFLECTION_OBJECT),

    new ObjectMemberFunctionCall("getValue",
      new ArrayAccess(REFLECTION_CLASS_GET_PROPERTIES),
      REFLECTION_CLASS_GET_PROPERTY,
      REFLECTION_PROPERTY),

    new ObjectMemberFunctionCall("newInstanceWithoutConstructor", REFLECTION_CLASS, REFLECTION_OBJECT),

    new ObjectMemberFunctionCall("setValue",
      new ArrayAccess(REFLECTION_CLASS_GET_PROPERTIES),
      REFLECTION_CLASS_GET_PROPERTY,
      REFLECTION_PROPERTY));

  private static final Predicate<TreeValues> UPDATE_PREDICATES = new ObjectMemberFunctionCall("setAccessible",
    new ArrayAccess(REFLECTION_CLASS_GET_METHODS, REFLECTION_CLASS_GET_PROPERTIES),
    REFLECTION_CLASS_GET_CONSTRUCTOR,
    REFLECTION_CLASS_GET_METHOD,
    REFLECTION_CLASS_GET_PROPERTY,
    REFLECTION_METHOD,
    REFLECTION_PROPERTY);

  @Override
  public void visitFunctionCall(FunctionCallTree tree) {
    TreeValues possibleValues = TreeValues.of(tree, context().symbolTable());
    if (BYPASS_PREDICATES.test(possibleValues)) {
      context().newIssue(this, tree, BYPASS_MESSAGE);
    } else if (UPDATE_PREDICATES.test(possibleValues)) {
      context().newIssue(this, tree, UPDATE_MESSAGE);
    }
    super.visitFunctionCall(tree);
  }

}
